internal/controller/servermaintenance_controller.go

```diff
@@ -150,59 +167,88 @@ func (r *ServerMaintenanceReconciler) handlePendingState(ctx context.Context, lo
-			return ctrl.Result{}, err
-		}
+		return r.moveToPrepareMaintenanceState(ctx, log, serverMaintenance, server)
 	}
 	if serverMaintenance.Spec.Policy == metalv1alpha1.ServerMaintenancePolicyEnforced {
 		log.V(1).Info("Enforcing maintenance", "Server", server.Name, "Maintenance", serverMaintenance.Name)
-		if err = r.updateServerRef(ctx, log, serverMaintenance, server); err != nil {
-			log.Error(err, "failed to patch server maintenance ref")
-			return ctrl.Result{}, err
-		}
-		if modified, err := r.patchMaintenanceState(ctx, serverMaintenance, metalv1alpha1.ServerMaintenanceStateInMaintenance); err != nil || modified {
-			return ctrl.Result{}, err
-		}
+		return r.moveToPrepareMaintenanceState(ctx, log, serverMaintenance, server)
 	}
 	return ctrl.Result{}, nil
 }
 
-func (r *ServerMaintenanceReconciler) handleInMaintenanceState(ctx context.Context, log logr.Logger, serverMaintenance *metalv1alpha1.ServerMaintenance) (ctrl.Result, error) {
+func (r *ServerMaintenanceReconciler) handlePrepareMaintenanceState(ctx context.Context, log logr.Logger, serverMaintenance *metalv1alpha1.ServerMaintenance) (ctrl.Result, error) {
 	server, err := r.getServerRef(ctx, serverMaintenance)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	config, err := r.applyServerBootConfiguration(ctx, log, serverMaintenance, server)
-	if err != nil {
-		return ctrl.Result{}, err
+
+	// rest of the operation can be done only if the server is in maintenance state
+	// else, we will hit conflict with ServerClaim controller which will still be asserting Specs on Serevr
+	if server.Status.State != metalv1alpha1.ServerStateMaintenance {
+		log.V(1).Info("Waiting for server to be in maintenance state", "Server", server.Name, "CurrentState", server.Status.State)
+		return ctrl.Result{}, nil
 	}
-	if config == nil {
-		if err := r.setAndPatchServerPowerState(ctx, log, server, serverMaintenance); err != nil {
+
+	var config *metalv1alpha1.ServerBootConfiguration
+	changed := false
+	if changed, config, err = r.isBootConfigurationChanged(ctx, log, serverMaintenance); err != nil {
+		return ctrl.Result{}, err
+	} else if changed || config == nil {
+		// turn server power off before applying new boot configuration
+		// this will help schedule the job to change the boot order for the server
+		// note: if the server is already off, this will be a no-op
+		if server.Spec.Power != metalv1alpha1.PowerOff && server.Status.PowerState != metalv1alpha1.ServerOffPowerState {
+			if err := r.setAndPatchServerPowerState(ctx, log, server, metalv1alpha1.PowerOff); err != nil {
+				return ctrl.Result{}, err
+			}
+			// requeue to give time to the server to power off
+			return ctrl.Result{RequeueAfter: r.ResyncInterval}, nil
+		}
+		// if no config found or if it has changed, we need to create or path it
+		config, err = r.applyServerBootConfiguration(ctx, log, serverMaintenance, server)
```

Comment:
Does this PR cover cover the use case to change bios settings, which don't need a reboot? In that case the boot config dance isn't strictly needed.